### DISCUSSION TITLE
remove_half_open was called unintentionally to introduce negative half_open counter?

### DISCUSSION THREAD

**Discussion by jzhang2000-g** [strongswan/strongswan#2403] [General]
Hi, we are using the strongswan 5.9.14 code, within checkout_by_message:
Suppose there was a first IKE_SA_INIT request #0 message, and was processed and put to half_open already. And then a retransmit of IKE_SA_INIT request #0 comes. 
`
METHOD(ike_sa_manager_t, checkout_by_message, ike_sa_t*,
	private_ike_sa_manager_t* this, message_t *message)
{
...
	if (message->get_request(message) &&
		message->get_exchange_type(message) == IKE_SA_INIT)
	{
		untrack_half_open = TRUE;     **<<< retransmit hits**
		if (message->get_message_id(message) == 0 &&
			id->get_responder_spi(id) == 0)
		{
			ike_version = IKEV2;   **<<<<<<< retransmit hits**
			is_init = TRUE;    
		}
	}
...
	if (is_init)
	{
...
		hasher = lib->crypto->create_hasher(lib->crypto, HASH_SHA1);  **<<<<< retransmit hits here**
		if (!hasher || !get_init_hash(hasher, message, &hash))
		{
			DBG1(DBG_MGR, "ignoring message, failed to hash message");
			DESTROY_IF(hasher);
			id->destroy(id);
			goto out;
		}
		hasher->destroy(hasher);

		/* ensure this is not a retransmit of an already handled init message */
		switch (check_and_put_init_hash(this, hash, &our_spi))   **<<<<< retransmit hits and found** 
		{
			case NOT_FOUND:
			{	...

					ike_sa = ike_sa_create(id, FALSE, ike_version);   **<<<<< Original IKE_SA_INIT request #0 hits **
					if (ike_sa)
					{
						entry = entry_create();
...
						entry->half_open = TRUE;
						entry->other = message->get_source(message);
						entry->other = entry->other->clone(entry->other);
						untrack_half_open = FALSE;

...						goto out;
					}

				remove_init_hash(this, hash);
				chunk_free(&hash);
				id->destroy(id);
				goto out;
			}
			case FAILED:
			{	/* we failed to allocate an SPI */
				chunk_free(&hash);
				id->destroy(id);
				DBG1(DBG_MGR, "ignoring message, failed to allocate SPI");
				goto out;
			}
			case ALREADY_DONE:
			default:
				break;
		}

		/* it looks like we already handled this init message to some degree */. **<<<<<<<< retransmit hits**
		id->set_responder_spi(id, our_spi);
		chunk_free(&hash);
		/* untrack the duplicate before waiting for the checkout */
		remove_half_open(this, message->get_source(message), FALSE);   **<<<<<<< Why we call remove_half_open() for the retransmit? it will decrease the half_open counters such as half_open_count_responder**
		untrack_half_open = FALSE;
	}

...
out:
	if (untrack_half_open)
	{
		remove_half_open(this, message->get_source(message), FALSE);
	}
...
	return ike_sa;
}
`

**Comment by strongX509** [comment]
Could you please state more clearly the actual problem that you are encountering.

**Comment by jzhang2000-g** [comment]
Hello,

Sorry for my late response, in international travel.

I have init_half_open_limit set as 200.

1) Peer sends me an IKE_SA_INIT request, OK, my half_open_count_responder
becomes 1, since 1 < 200 (the limit), everything is fine.
2) Peer sends me a retransmit of this IKE_SA_INIT request, due to this
suspect extra remove_half_open call, my half_open_count_responder gets
decreased to 0.
3) My strongswan processes the IKE_SA_INIT request received in step 1), and
advances the negotiation, and we need to call remove_half_open again, then
half_open_count_responder becomes 0 - 1 = -1 (4294967295)
4) For any future negotiation attempts from any peer, since my
half_open_count_responder (4294967295) is already much bigger than
init_half_open_limit (200), my strongswan is already stuck!, not able to do
any new negotiations

Please double check

Thanks,
Jimmy

**Reply by tobiasbrunner** [reply]
> Peer sends me a retransmit of this IKE_SA_INIT request

Which increases the counter:

https://github.com/strongswan/strongswan/blob/12d2b6e2b4a0b33be4f69250e20d071ab1a184ca/src/libcharon/processing/jobs/process_message_job.c#L137-L142

**Reply by jzhang2000-g** [reply]
Thanks. It turns out to be our own change. We have bumped half_open_count_responder much sooner, (basically as soon as the entry was created), instead of waiting for put_half_open to be called to bump the half_open_count_responder.
In extreme condition, when the first SA negotiation is done and put_half_open is called, there are already hundreds of SA entry created and managed to sneak into the negotiation queue, which breaks the init_half_open_limit condition we set.

**Reply by tobiasbrunner** [reply]
I don't get what you mean, please explain.

**Reply by jzhang2000-g** [reply]
Basically we had some of our own fix in earlier strongswan 5.9.2 code to bump the half_open counters sooner. Now 5.9.14 has a native fix (by introducing track_init, i.e. put_half_open to process_message_job.c to bump the half_open_counters earlier)

In strongswan 5.9.2, put_half_open is called after negotiation is done, it is already too late if we want to use init_half_open_limit.
In strongswan 5.9.14, put_half_open is called before the negotiation (in process_message_job.c), this fix already addresses our original issue nicely. Thus we are going to take our own fix out, no need anymore.

**Reply by tobiasbrunner** [reply]
Ah, OK. Thanks for the explanation.
